### PR TITLE
Constants and variables

### PR DESCRIPTION
**Pull request by SamFerracin**
**Context:**
Improving variables handling, in preparation for the backend refactoring and the Fock-Bargman refactoring

**Description of the Change:**
- Removed `parameters.py` and `parametrized.py`, with all their functions and methods. Gates, states, and detectors no longer inherit from `Parametrized`.
- Added `Constant` and `Variable` classes to represent constants and variables, and `ParameterSet` to represent sets of parameters. Gates, states, and detectors own `ParameterSet`s and have fast access to constant and variable parameters.

**NOT PART OF THIS PR:**
Improvements to the training package (not priority)

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -26,6 +26,10 @@ This is done by representing Fock amplitudes with a higher precision than comple
 to keep the code fast.
   [(#274)](https://github.com/XanaduAI/MrMustard/pull/274)
 
+* Replaced parameters in `training` with `Constant` and `Variable` classes.
+
```
Comment (ziofil) [inline .github/CHANGELOG.md:30]: please add link to PR

File: mrmustard/lab/gates.py [hunk 2]
```diff
@@ -81,24 +83,15 @@ def __init__(
...
-            self,
-            x=x,
-            y=y,
-            x_trainable=x_trainable,
-            y_trainable=y_trainable,
-            x_bounds=x_bounds,
-            y_bounds=y_bounds,
-            **kwargs,
-        )
         Unitary.__init__(
```
Comment (ziofil) [inline mrmustard/lab/gates.py:88]:
```suggestion
        super().__init__(modes=modes or list(range(m)), name="Dgate")
```

Comment (ziofil) [inline mrmustard/lab/gates.py:88]:
```suggestion
        super().__init__(modes=modes or list(range(m)), name="Dgate")
```

File: mrmustard/lab/states.py [hunk 3]
```diff
@@ -439,21 +409,19 @@ def __init__(
...
         self._modes = modes
         self._normalize = normalize
 
-        cov = gaussian.gaussian_cov(self.symplectic.value, self.eigenvalues.value)
+        self._parameter_set = ParameterSet()
+        eb = (settings.HBAR / 2, None) if eigenvalues_bounds == (None, None) else eigenvalues_bounds
+        self._add_parameter(make_parameter(eigenvalues_trainable, eigenvalues, "eigenvalues", eb))
+        self._add_parameter(
+            make_parameter(
+                symplectic_trainable, symplectic, "symplectic", (None, None), update_symplectic
```
Comment (ziofil) [inline mrmustard/lab/states.py:420]: very nice 👍

File: mrmustard/math/parameters.py [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        """
+        return self._value
+
+    @value.setter
+    def value(self, value):
+        self._value = self._get_value(value, self.bounds, self.name)
+
+    @staticmethod
+    def orthogonal(
+        value: any,
```
Comment (ziofil) [inline mrmustard/math/parameters.py:206]: if value=None, can we initialize a random one? We have the code already in math. Same for unitary and symplectic

File: mrmustard/math/parameters.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+            bounds: The numerical bounds of the returned variable.
+
+        Returns:
+            A variable with ``update_fn`` for unitary optimization.
+        """
+        return Variable(value, name, bounds, update_unitary)
+
+    def __mul__(self, value):
+        return type(self)(
+            value=value * self.value,
```
Comment (ziofil) [inline mrmustard/math/parameters.py:263]: why with `*`?

**Review by ziofil — APPROVED** [review]
Super!

File: mrmustard/lab/gates.py (hunk 2 above)
Comment (SamFerracin) [reply mrmustard/lab/gates.py:88]: Thanks, I changed them all

**Pull request merged**